File: src/main/java/architecture/goldenboughs_lib/datagen/LibDatagenParticle.java
package architecture.goldenboughs_lib.datagen;

import architecture.goldenboughs_lib.client.particle.LcDamageIconParticle;
import architecture.goldenboughs_lib.core.Lib;
import architecture.goldenboughs_lib.init.LibParticleTypes;
import net.minecraft.core.particles.ParticleType;
import net.minecraft.data.PackOutput;
import net.neoforged.neoforge.common.data.ExistingFileHelper;
import net.neoforged.neoforge.common.data.ParticleDescriptionProvider;

import java.util.Arrays;
import java.util.function.Supplier;
import java.util.stream.Collectors;

public final class LibDatagenParticle extends ParticleDescriptionProvider {
	public LibDatagenParticle(PackOutput output, ExistingFileHelper fileHelper) {
		super(output, fileHelper);
	}

	@Override
	protected void addDescriptions() {
		sprite(LibParticleTypes.LC_DAMAGE_ICON, Arrays.stream(LcDamageIconParticle.Type.values())
			.map(LcDamageIconParticle.Type::getTexturePl)
			.toArray(String[]::new));
	}

	private <T extends ParticleType<?>> void sprite(Supplier<T> type, String name) {
		sprite(type.get(), Lib.modRl(name));
	}

	private <T extends ParticleType<?>> void sprite(Supplier<T> type, String... names) {
		spriteSet(type.get(), Arrays.stream(names)
			.map(Lib::modRl)
			.collect(Collectors.toList()));
	}
}

File: src/main/kotlin/architecture/goldenboughs_lib/datagen/LibDatagenParticle.kt
package architecture.goldenboughs_lib.datagen

import architecture.goldenboughs_lib.core.LibConstants.modRl
import architecture.goldenboughs_lib.init.LibParticleTypes
import architecture.goldenboughs_lib.module.lc_damage.particle.LcDamageIconParticle
import net.minecraft.core.particles.ParticleType
import net.minecraft.data.PackOutput
import net.neoforged.neoforge.common.data.ExistingFileHelper
import net.neoforged.neoforge.common.data.ParticleDescriptionProvider
import java.util.function.Supplier

class LibDatagenParticle(output: PackOutput, fileHelper: ExistingFileHelper) :
	ParticleDescriptionProvider(output, fileHelper) {
	override fun addDescriptions() {
		sprite(
			LibParticleTypes.LC_DAMAGE_ICON,
			*LcDamageIconParticle.Type.entries.map { it.texturePl }.toTypedArray()
		)
	}

	private fun <T : ParticleType<*>> sprite(type: Supplier<T>, name: String) {
		sprite(type.get(), modRl(name))
	}

	private fun <T : ParticleType<*>> sprite(type: Supplier<T>, vararg names: String) {
		spriteSet(type.get(), names.map { modRl(it) })
	}
}
